dhis-2/dhis-api/src/main/java/org/hisp/dhis/program/notification/ProgramNotificationInstanceService.java

```diff
@@ -47,4 +47,6 @@ List<ProgramNotificationInstance> getProgramNotificationInstances(
   Long countProgramNotificationInstances(ProgramNotificationInstanceParam params);
 
   void validateQueryParameters(ProgramNotificationInstanceParam param);
+
+  List<ProgramNotificationInstance> getAll();
```

Comment:
I agree with Ivo, this was the major culprit of the performance issue and it is only used in tests.
Can we remove it?

Reply:
The main performance issue was caused by `ProgramNotificationInstanceStore.getAll()`.
To test this properly, I added a corresponding method in `ProgramNotificationInstanceService`, which is now used in `ProgramNotificationInstanceServiceTest`.

Reply:
> The main performance issue was caused by ProgramNotificationInstanceStore.getAll().

which you then fixed by removing the orgUnit snapshot and using 
 `notificationInstanceService.getProgramNotificationInstances(param)` to get only the instances scheduled for today. This means that you should test `notificationInstanceService.getProgramNotificationInstances(param)` returns the instances you would expect (does so in a reasonable amount of time). If `ProgramNotificationInstanceStore.getAll()` is not used anymore you should remove it.

Reply:
the performance fix involved two distinct changes:

- Filtering notifications in the database instead of in-memory (getProgramNotificationInstances(param)).
- Reducing the JSONB payload (by excluding unnecessary data like organisationUnit from the snapshot).

Both changes should be tested isn't? If not so, then we can remove `.getAll()`

Reply:
> Filtering notifications in the database instead of in-memory (getProgramNotificationInstances(param)).
For the above change, as long as you test/assert that the method is filtering notifications in the database and not in-memory, that should suffice. You don't necessarily need to test the "performance" side of this. 

> Reducing the JSONB payload (by excluding unnecessary data like organisationUnit from the snapshot).
For the above change, maybe simply asserting organisationUnit is not part of the snapshot json may also suffice? Again no need to create integration tests that focus on performance for this I feel. 

Thoughts @zubaira  @enricocolasante @teleivo ?

Reply:
I discussed this with @zubaira. We settled on writing an integration test for the service that gets the notification instances of today. That previously would have timed out. So now having a test for it that finishes is proof enough. We now have a global Junit timeout on tests of 1m so there is not even a need to add a `@Timeout` annotation on this test.

So even though the description mentions performance we test that it works without any target numbers.